### ISSUE TITLE
Import behaviour if no Qt bindings are installed

### ISSUE THREAD

**Issue by tgwoodcock** [spyder-ide/qtpy#367]
Hi, firstly thanks very much for QtPy - the flexibility it brings has been very useful for a recent Qt gui I've been working on. While testing that project, I noticed something which I think is unintended behaviour with qtpy.

This concerns only the case where *no Qt bindings are installed*. In a clean venv, with only qtpy, pip, wheel and setuptools installed, with the current version of qtpy, 

```python
>>> import qtpy
>>> qtpy.API_NAME
'PyQt5'
```

importing the package succeeds and the API_NAME is set to "PyQt5", even though there are no Qt bindings installed. Also e.g.:

```python
>>> import qtpy.QtCore
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "E:\test_qtpy\qtpy\qtpy\QtCore.py", line 15, in <module>
    from PyQt5.QtCore import *
ModuleNotFoundError: No module named 'PyQt5'
```

results in a ModuleNotFoundError because the program is looking for PyQt5, rather than raising a QtBindingsNotFoundError.

- I think that this can be fixed by replacing line 259 in \_\_init\_\_.py with "raise QtBindingsNotFoundError"

- There also appears to be a bug in the definition of QtBindingsNotFoundError: in line 84 in \_\_init\_\_.py, a "self." needs to be inserted before _msg.

With these two changes:

```python
>>> import qtpy
Traceback (most recent call last):
  File "E:\test_qtpy\lib\site-packages\qtpy\__init__.py", line 252, in <module>
    from PySide6 import __version__ as PYSIDE_VERSION  # analysis:ignore
ModuleNotFoundError: No module named 'PySide6'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "E:\test_qtpy\lib\site-packages\qtpy\__init__.py", line 259, in <module>
    raise QtBindingsNotFoundError
qtpy.QtBindingsNotFoundError: No Qt bindings could be found
```

- As \_\_init\_\_.py is executed before importing any other modules, an attempted import of any other module will result in the QtBindingsNotFoundError
- I think this means it may be unnecessary for other modules to import the QtBindingsNotFoundError, as this case would be already handled by \_\_init\_\_.py.

Hope this might be useful. Thanks again.

**Comment by CAM-Gerlach** [comment]
Hey @tgwoodcock , thanks for your detailed report!

IIRC, there are historical reasons for falling back to blindly trying PyQt5 if Qt bindings cannot be found, but I'm not sure how relevant they still are now that QtPy is used far beyond Spyder, which originally only really worked well with PyQt5. It would seem cleaner and more sensible, at least starting from a clean slate, to behave more like you suggest. There may, however, still be some some practical reasons for the current behavior of naive fallback, aside from just making a few package-level constants and helper functions available even if no Qt binding is installed — @dalthviz , any insight here?

Besides that, this does raise significant backward compatibility concerns with existing applications that expect and rely on the current handling when no bindings are found, since it both raises an error at a different time (upon importing `qtpy`, which applications may import first to e.g. check global constants, versions or other reasons, rather than just most submodules) and also raises a different error in many cases (since with a PyQt5 default fallback, the logic in many of the submodules will raise a `QtModuleNotInstalledError`, `QtModuleNotInOSError`, `QtModuleNotInQtVersionError` or `QtBindingMissingModuleError` rather than a `QtBindingsNotFoundError`). I'd argue the latter is actually a bug we overlooked, or at least undesirable and confusing behavior, and at least in many cases was introduced fairly recently with #344 , but still means we should consider a major version bump for this per SemVer (or at the very least another minor version bump).

However, if we do that, it would also be a great opportunity to refine and refactor the binding selection behavior more generally (perhaps along the lines of the ideas in the stale #156 ), as it currently has a hardcoded binding try order that cannot be changed by clients, except to force it to start at a specific binding/version combination, but that means any bindings earlier in the list are not tried at all. It also has a fair bit of duplication and code smell and can be hard to follow and reason with, which that would resolve.

> * I think that this can be fixed by replacing line 259 in __init__.py with "raise QtBindingsNotFoundError"

Yup, that does indeed appear to be a straightforward bug, thanks.

> importing the package succeeds and the API_NAME is set to "PyQt5", even though there are no Qt bindings installed.

This raises a good point—right now, `API_NAME` indicates what binding QtPy will (naively or otherwise) try, but there isn't a good way other than EAFP to determine whether a Qt binding was found or QtPy fell back to its naive guess. The behavior change you propose (plus some `__init__.py` refactoring) would obviate the need for it, but if we defer that to 3.x due to compat concerns and want a stopgap for now, it would probably be a good idea to at least introduce a new global constant to indicate whether (and perhaps even how) a binding was found.

> I think that this can be fixed by replacing line 259 in __init__.py with "raise QtBindingsNotFoundError"

A few issues with this specific implementation—First, the error isn't raised with `raise from`, so it appears as an unhandled bug in the exception-handling code itself, rather than the `ModuleNotFoundError` importing the (last) binding being the deliberate _cause_ of this error (i.e. the latter wrapping the former). Second, the broader issue, at least with the current (decidedly suboptimal) `__init__.py` behavior, is that the error gets raised inside the PySide6 branch, which will quite possibly confuse users into thinking QtPy (only) tried to use that binding and not that it simply arbitrarily happened to be the last fallback it tried.

Therefore, to avoid this in the current implementation, we could simply do `raise QtBindingsNotFoundError from None` which would avoid this potentially confusing exception context and/or cause. Of course, if we refactored the binding try order logic, we would likely avoid the need for this.

**Comment by CAM-Gerlach** [comment]
> I'm not sure to be honest, maybe @ccordoba12 has a better idea if the current behavior is intentional or not. However, checking the behavior with QtPy 1.11.3 you can't `import qtpy` it raises a `PythonQtError('No Qt bindings could be found')` while with QtPy 2.0.0 `import qtpy` works without rasing an error (I'm not totally sure if at somepoint for 2.x we decided to enable the import of qtpy without bindings available thinking).

Ah, great find, @dalthviz . From digging into the [Git history](https://github.com/spyder-ide/qtpy/blame/cc9086e596ccbeb63913a8c245c5c8dd2c9d0c12/qtpy/__init__.py#L191) it looks like this was an unintentional behavior change that was overlooked in the Qt4 + 5 to Qt5 + 6 transition. There isn't any commits where this was deliberately added, and instead what happened is the old Pyside `if` block, previously the last in the sequence which contained the `raise PythonQtError`, was deleted, and the PySide6 code block copy/pasted from the previous PySide2 code block that had its fallback to the next in the fixed sequence, `pyqt` (i.e. PyQt4), which we dropped support for, which was not spotted at the time.

Months later, when that bug *was* spotted, it was assumed to be an intentional fallback rather than being a simple copy/paste mistake, and later changes to that file (including mine) went with that assumption. However, in reality, this "fallback" behavior was simply an oversight in QtPy 2.0 that was never conclusively spotted until now. Of course, it is easy to make these sorts of logic errors given the current structure of the checks in the `__init__.py`, which speaks to the benefit in us refactoring them.

Therefore, the previous and intended behavior was exactly as @tgwoodcock suggestions (with the issues and refinements I bring up above re the exception chaining). Therefore, the bug should be fixed as soon as practical, as it is a clear defect and not intended behavior. However, it is worth considering the possible impact of what is still a behavior change for clients that have adapted to it. In particular, it would require the following conditions to become effective:

* Application depends on QtPy >=2
* User does not have any supported binding installed
* Application imports a QtPy submodule in a try-except block, and EITHER:
        * `except` block checks for the new (in QtPy 2.2.0) `QtModuleNotFoundError` or a subclass rather than just the parent `PythonQtError`
        * Application also imports the top-level `qtpy` package elsewhere without a try-except, that runs before any protected submodule import
        
This seems like a fairly niche corner case, but still worth at least a minor release and a clear note in the release notes/changelog of the change, and maybe taking a look at a handful of major dependents to see if that is the case.

> Set the `API_NAME` value to something like `not found` if not binding is found maybe?

In case we still do something like this for 2.x (i.e. still retain the fallback until 3.x) it would make more sense to simply set it to `None`, the canonical Python missing value, rather an arbitrary string.

**Comment by dalthviz** [comment]
So I think there are already a couple of things for v2.2.1, maybe we should fix the missing `self` of the `_msg` call on the `QtBindingsNotFoundError` definition and release v2.2.1 and then restore raising the exception (although now it will be `QtBindingsNotFoundError`) for v2.3.0?

What do you think @CAM-Gerlach ?  Do you see any advantage on keeping the current behavior?

**Comment by CAM-Gerlach** [comment]
Sounds like a good plan to me :+1: 

> Do you see any advantage on keeping the current behavior?

Nope, not really, other than theoretical backward compatibility with a likely narrow corner case (that is really an unintended bug). I do think at least for 3.0 we should refactor the `__init__` along the lines of a simplified #156 to avoid these sorts of issues in the future.
